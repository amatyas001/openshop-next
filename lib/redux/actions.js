export * from './cart';
export * from './filter';
export * from './panel';
export * from './payment';
