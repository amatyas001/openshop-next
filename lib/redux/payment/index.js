export * from './paymentCancel';
export * from './paymentCreate';
export * from './paymentStage';
