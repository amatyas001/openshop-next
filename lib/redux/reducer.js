import {
  CART_MANAGE,
  PAYMENT_STAGE,
  ITEM_FILTER,
  PANEL_TOGGLE,
} from '@app/lib/redux/actions';

export const reducer = (state, action) => {
  switch (action.type) {
    case PANEL_TOGGLE: {
      // Toggle UI panel collapse states
      return {
        ...state,
        panel: { ...state.panel, [action.panel.panel]: action.panel.value },
      };
    }
    case ITEM_FILTER: {
      // Merge item filters
      return {
        ...state,
        filters: { ...state.filters, ...action.filters },
      };
    }

    case CART_MANAGE:
      switch (action.cart.status) {
        case 'add':
          // Increase the amount of an existing item or add a new entry
          // to the beginning of the cart
          return {
            ...state,
            cart:
              [...state.cart].filter(
                (c) =>
                  c.id === action.cart.item.id &&
                  c.buy.size === action.cart.item.buy.size &&
                  c.buy.color === action.cart.item.buy.color
              ).length > 0
                ? [...state.cart].reduce((a, c) => {
                    a.push(
                      /* istanbul ignore next */
                      c.id === action.cart.item.id
                        ? {
                            ...c,
                            buy: {
                              ...c.buy,
                              amount:
                                c.buy.amount + action.cart.item.buy.amount,
                            },
                          }
                        : c
                    );
                    return a;
                  }, [])
                : [action.cart.item, ...state.cart],
          };

        case 'remove':
          // Decrease the amount of a cart item or skip it when amount is zero
          return {
            ...state,
            cart: [...state.cart].reduce((a, c) => {
              /* istanbul ignore next */
              c.id === action.cart.item.id
                ? c.buy.amount - 1 > 0 &&
                  a.push({
                    ...c,
                    buy: { ...c.buy, amount: c.buy.amount - 1 },
                  })
                : a.push(c);
              return a;
            }, []),
          };
      }

    case PAYMENT_STAGE: {
      // Put a new token if its not set and pass payment data to state
      return action.payment.status === 'review'
        ? state.payment && state.payment.token
          ? {
              ...state,
              payment: {
                ...state.payment,
                status: action.payment.status,
              },
            }
          : {
              ...state,
              payment: {
                ...action.payment,
              },
            }
        : {
            ...state,
            payment: {
              ...state.payment,
              ...action.payment,
            },
            cart: /* istanbul ignore next */ action.cart
              ? [...action.cart]
              : [...state.cart],
          };
    }

    default:
      return { ...state };
  }
};
