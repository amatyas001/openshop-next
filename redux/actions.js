export const ADD_TO_CART = 'ADD_TO_CART';
export const REMOVE_FROM_CART = 'REMOVE_FROM_CART';
export const CLEAR_CART = 'CLEAR_CART';
export const GET_AMOUNT = 'GET_AMOUNT';
export const CHECKOUT_PROGRESS = 'CHECKOUT_PROGRESS';
export const PAYMENT_PROGRESS = 'PAYMENT_PROGRESS';
export const PAYMENT_RESET = 'PAYMENT_RESET';

/**
 *  CART ACTIONS
 */
export function addToCart(item) {
  return {
    type: ADD_TO_CART,
    payload: item,
  };
}

export function removeFromCart(item) {
  return {
    type: REMOVE_FROM_CART,
    payload: item,
  };
}

export function clearCart() {
  return {
    type: CLEAR_CART,
  };
}

export function getAmount() {
  return {
    type: GET_AMOUNT,
  };
}

/**
 *  CHECKOUT FLOW
 */
export function checkoutReview() {
  return {
    type: CHECKOUT_PROGRESS,
    payload: { status: 'review' },
  };
}

export function checkoutDetails() {
  return {
    type: CHECKOUT_PROGRESS,
    payload: { status: 'details' },
  };
}

export function checkoutFinished() {
  return {
    type: CHECKOUT_PROGRESS,
    payload: { status: 'finished' },
  };
}

export function paymentActive() {
  return {
    type: PAYMENT_PROGRESS,
    payload: { status: true },
  };
}

export function paymentStart() {
  return {
    type: PAYMENT_PROGRESS,
    payload: { status: 'loading' },
  };
}

/**
 *  PAYMENT FLOW
 */
export function paymentConfirm(card, details, intent) {
  return {
    type: PAYMENT_PROGRESS,
    payload: {
      status: 'confirm',
      card,
      details,
      intent,
    },
  };
}

export function paymentSuccess(intent) {
  return {
    type: PAYMENT_PROGRESS,
    payload: {
      status: 'success',
      intent,
    },
  };
}

export function paymentCanceled(intent) {
  return {
    type: PAYMENT_PROGRESS,
    payload: {
      status: 'canceled',
      intent,
    },
  };
}

export function paymentError(error) {
  return {
    type: PAYMENT_PROGRESS,
    payload: { status: 'error', error },
  };
}

export function paymentReset() {
  return {
    type: PAYMENT_RESET,
  };
}

export function cancelPayment(id) {
  return async (dispatch) => {
    try {
      dispatch(paymentStart());
      const { loadStripe } = await import('@stripe/stripe-js');
      const stripe = await loadStripe(
        process.env.NEXT_PUBLIC_STRIPE_PUBLIC_KEY
      );
      stripe.paymentIntents.cancel(id, (err, intent) => {
        if (err) dispatch(paymentError(e.message));
        else dispatch(paymentCanceled(intent.id));
      });
    } catch (e) {
      dispatch(paymentError(e.message));
    }
  };
}

export function intentCart(card, details, items) {
  return async (dispatch) => {
    try {
      dispatch(paymentStart());
      const axios = await import('axios');
      const intent = await axios.post(
        process.env.NODE_ENV === 'production'
          ? 'https://openshop.netlify.com/.netlify/functions/intent'
          : 'http://localhost:9000/.netlify/functions/intent',
        {
          items,
        },
        {
          headers: { 'Content-Type': 'application/json' },
        }
      );
      dispatch(paymentConfirm(card, details, intent.data));
    } catch (e) {
      dispatch(paymentError(e.message));
    }
  };
}
