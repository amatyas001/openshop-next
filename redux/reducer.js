import {
  SCROLL,
  ADD_TO_CART,
  REMOVE_FROM_CART,
  CLEAR_CART,
  GET_AMOUNT,
  CHECKOUT_PROGRESS,
  PAYMENT_PROGRESS,
  PAYMENT_RESET,
  PAYMENT_TOKEN,
} from './actions';

export function reducer(state, action) {
  switch (action.type) {
    case SCROLL:
      return { ...state, scroll: action.payload };
    case ADD_TO_CART:
      return { ...state, cart: [...state.cart, action.payload] };
    case REMOVE_FROM_CART:
      return {
        ...state,
        cart: [...state.cart].filter((item) => item.id !== action.payload.id),
      };
    case CLEAR_CART:
      return { ...state, cart: [], checkout: 'success' };
    case GET_AMOUNT:
      return {
        ...state,
        amount:
          state.cart &&
          state.cart.reduce((a, c) => (a += parseFloat(c.price)), 0).toFixed(2),
      };
    case CHECKOUT_PROGRESS: {
      return {
        ...state,
        checkout: action.payload,
      };
    }
    case PAYMENT_PROGRESS: {
      return {
        ...state,
        payment: action.payload,
        checkout:
          action.payload.status === true
            ? { status: 'review' }
            : state.checkout,
      };
    }
    case PAYMENT_RESET: {
      return {
        ...state,
        checkout: { status: 'review' },
        payment: { status: 'reset' },
      };
    }
    case PAYMENT_TOKEN: {
      if (state.payment.token) return { ...state };
      return {
        ...state,
        payment: { ...state.payment, token: action.payload },
      };
    }
    default:
      return { ...state };
  }
}
