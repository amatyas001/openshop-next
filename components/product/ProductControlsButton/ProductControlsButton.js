import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import * as COLORS from '@app/config/colors';
import { addToCart } from '@app/lib/redux/actions';
import { Button } from '@app/components';
import { ProductShape } from '@app/lib/types';

// Helper function to check the validity of given shopping details
const valid = (state) =>
  (!Array.isArray(state.product.color)
    ? state.product.color
    : state.product.buy.color) &&
  (state.product.sizes ? state.product.buy.size : 'onesize') &&
  state.product.buy.amount > 0;

/**
 * Renders a button to add a specific product to the cart.
 *
 * @see https://amatyas001.github.io/openshop-next/#productcard
 */
export const ProductControlsButton = React.memo(
  (props) => {
    const { product, setDetails } = props;
    const dispatch = useDispatch();

    return (
      <Button
        onClick={() => {
          dispatch(addToCart(product));
          setDetails({ ...product, buy: { amount: 0 } });
        }}
        height='50px'
        disabled={!valid(props)}
        data-testid='product-buy-button'
        _disabled={{
          ...COLORS.BUTTON.secondary.hover,
        }}
        {...props}
      >
        buy
      </Button>
    );
  },
  /* istanbul ignore next */
  (prev, next) => valid(prev) === valid(next)
);

ProductControlsButton.defaultProps = {
  product: {
    buy: {
      amount: 0,
    },
  },
};

ProductControlsButton.propTypes = {
  /**
   * Unique item object with buying details
   */
  product: PropTypes.shape({
    ...ProductShape,
    buy: PropTypes.shape({
      amount: PropTypes.number.isRequired,
      color: PropTypes.string,
      size: PropTypes.string,
    }),
  }).isRequired,
};
