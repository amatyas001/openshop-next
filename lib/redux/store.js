import { useMemo } from 'react';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { reducer } from '@app/lib/redux/reducer';

export const initStore = (reducers, initialState) => {
  return createStore(reducers, initialState, applyMiddleware(thunkMiddleware));
};

export const initializeStore = (preloadedState) => {
  let store;

  if (typeof window !== 'undefined') {
    store = initStore(
      persistReducer(
        {
          key: 'openshop',
          storage,
        },
        reducer
      ),
      {
        ...preloadedState,
      }
    );
    store.__PERSISTOR = persistStore(store);
  } else {
    store = initStore(reducer, preloadedState);
  }

  return store;
};

export function useStore(initialState) {
  const store = useMemo(() => initializeStore(initialState), [initialState]);
  return store;
}
