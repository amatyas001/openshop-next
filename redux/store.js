import { useMemo } from 'react';
import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { reducer } from './reducer';

const persisted = persistReducer(
  {
    key: 'cart',
    storage,
  },
  reducer
);

function initStore(reducers, initialState) {
  return createStore(reducers, initialState, applyMiddleware(thunkMiddleware));
}

export const initializeStore = (preloadedState) => {
  let store;
  if (typeof window !== 'undefined') {
    store = initStore(persisted, {
      ...preloadedState,
    });
    store.__PERSISTOR = persistStore(store);
  } else {
    store = initStore(reducer, preloadedState);
  }
  return store;
};

export function useStore(initialState) {
  const store = useMemo(() => initializeStore(initialState), [initialState]);
  return store;
}
