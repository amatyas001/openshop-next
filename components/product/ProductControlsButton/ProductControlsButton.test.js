import { create, act } from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { ProductControlsButton } from './ProductControlsButton';
import { addToCart } from '@app/lib/redux/actions';

const mockStore = configureStore([]);

const mock_handler = jest.fn();

describe('<ProductControlsButton />', () => {
  let tree, store, details;

  beforeAll(() => {
    details = {
      id: 'mock_id',
      name: 'mock_name',
      price: 10,
      amount: 10,
      color: 'mock_color',
      starrating: 3,
      buy: {
        amount: 0,
      },
    };

    act(() => {
      tree = create(
        <Provider store={mockStore({})}>
          <ProductControlsButton product={details} setDetails={mock_handler} />
        </Provider>
      );
    });
  });

  it('should render with required props', () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should be disabled if details are invalid', () => {
    expect(
      tree.root.findByProps({
        'data-testid': 'product-buy-button',
      }).props.disabled
    ).toBeTruthy();
  });

  let button,
    config = [
      {
        name: 'when color and size is static',
        details: {
          id: 'mock_id',
          name: 'mock_name',
          price: 10,
          color: 'mock_color',
          amount: 10,
          starrating: 3,
          buy: {
            amount: 1,
          },
        },
      },
      {
        name: 'when color is array',
        details: {
          id: 'mock_id',
          name: 'mock_name',
          price: 10,
          color: ['mock_color'],
          size: 'mock_size',
          amount: 10,
          starrating: 3,
          buy: {
            color: 'mock_color',
            amount: 1,
          },
        },
      },
      {
        name: 'when size is array',
        details: {
          id: 'mock_id',
          name: 'mock_name',
          price: 10,
          color: 'mock_color',
          size: ['mock_size'],
          amount: 10,
          starrating: 3,
          buy: {
            amount: 1,
            size: 'mock_size',
          },
        },
      },
      {
        name: 'when color and size are arrays',
        details: {
          id: 'mock_id',
          name: 'mock_name',
          price: 10,
          color: ['mock_color'],
          size: ['mock_size'],
          amount: 10,
          starrating: 3,
          buy: {
            amount: 1,
            size: 'mock_size',
            color: 'mock_color',
          },
        },
      },
    ];

  config.forEach((item) => {
    describe('with valid details', () => {
      beforeAll(() => {
        store = mockStore({});

        act(() => {
          tree = create(
            <Provider store={store}>
              <ProductControlsButton
                product={item.details}
                setDetails={mock_handler}
              />
            </Provider>
          );
        });

        button = tree.root.findByProps({
          'data-testid': 'product-buy-button',
        });
      });

      afterAll(() => {
        mock_handler.mockClear();
      });

      it(`should be enabled ${item.name}`, () => {
        expect(button.props.disabled).toBeFalsy();
      });

      it(`should dispatch addToCart action ${item.name}`, () => {
        act(() => {
          button.props.onClick();
        });

        expect(store.getActions()).toEqual([addToCart(item.details)]);
      });

      it(`should reset details ${item.name}`, () => {
        expect(mock_handler).toHaveBeenCalledTimes(1);
        expect(mock_handler).toHaveBeenCalledWith({
          ...item.details,
          buy: {
            amount: 0,
          },
        });
      });
    });
  });
});
