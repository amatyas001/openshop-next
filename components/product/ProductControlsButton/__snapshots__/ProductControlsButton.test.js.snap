// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductControlsButton /> should render with required props 1`] = `
<button
  className="css-1e5lh77"
  data-testid="product-buy-button"
  disabled={true}
  href="#void"
  onClick={[Function]}
  onKeyDown={[Function]}
  tabIndex="0"
  type="button"
>
  buy
</button>
`;
